peer.go

```diff
@@ -74,81 +83,76 @@ func (l *PeerList) newSibling() *PeerList {
 
 // Add adds a peer to the list if it does not exist, or returns any existing peer.
 func (l *PeerList) Add(hostPort string) *Peer {
-	l.RLock()
-
-	if p, ok := l.peersByHostPort[hostPort]; ok {
-		l.RUnlock()
-		return p
+	if ps, ok := l.exists(hostPort); ok {
+		return ps.Peer
 	}
-
-	l.RUnlock()
 	l.Lock()
 	defer l.Unlock()
 
 	if p, ok := l.peersByHostPort[hostPort]; ok {
-		return p
+		return p.Peer
 	}
 
 	p := l.parent.Add(hostPort)
-	atomic.AddUint32(&p.scCount, 1)
-	l.peersByHostPort[hostPort] = p
-	l.peers = append(l.peers, p)
+	ps := newPeerScore(p)
+	ps.score = l.scoreCalculator.GetScore(p)
+
+	l.peersByHostPort[hostPort] = ps
+	l.peerHeap.PushPeer(ps)
 	return p
 }
 
 // Get returns a peer from the peer list, or nil if none can be found.
-// Get will avoid peers in prevSelected unless there are no other peers left.
 func (l *PeerList) Get(prevSelected map[string]struct{}) (*Peer, error) {
-	l.RLock()
-
-	if len(l.peers) == 0 {
-		l.RUnlock()
+	l.Lock()
```

Comment:
I don't have it right now. It starts using lock because the choosePeer will pop up peer because of retry.